backend/dataall/modules/worksheets/services/worksheet_query_result_service.py

```diff
@@ -0,0 +1,121 @@
+import csv
+import io
+import os
+from datetime import datetime, timedelta
+from typing import TYPE_CHECKING
+
+from openpyxl import Workbook
+
+from dataall.base.db import exceptions
+from dataall.core.environment.services.environment_service import EnvironmentService
+from dataall.core.permissions.services.tenant_policy_service import TenantPolicyService
+from dataall.modules.worksheets.aws.s3_client import S3Client
+from dataall.modules.worksheets.db.worksheet_models import WorksheetQueryResult
+from dataall.modules.worksheets.db.worksheet_repositories import WorksheetRepository
+from dataall.modules.worksheets.services.worksheet_enums import WorksheetResultsFormat
+from dataall.modules.worksheets.services.worksheet_permissions import RUN_ATHENA_QUERY_TENANT
+from dataall.modules.worksheets.services.worksheet_service import WorksheetService
+
+if TYPE_CHECKING:
+    try:
+        from sqlalchemy.orm import Session
+        from openpyxl.worksheet.worksheet import Worksheet
+    except ImportError:
+        print('skipping type checks as stubs are not installed')
+        Session = None
+        Worksheet = None
+
+
+class WorksheetQueryResultService:
+    _DEFAULT_ATHENA_QUERIES_PATH = 'athenaqueries'
+    _DEFAULT_QUERY_RESULTS_TIMEOUT = os.getenv('QUERY_RESULT_TIMEOUT_MINUTES', 120)
+
+    @staticmethod
+    def _create_query_result(
+        environment_bucket: str, athena_workgroup: str, worksheet_uri: str, region: str, aws_account_id: str, data: dict
+    ) -> WorksheetQueryResult:
+        sql_query_result = WorksheetQueryResult(
+            worksheetUri=worksheet_uri,
+            AthenaQueryId=data.get('athenaQueryId'),
+            fileFormat=data.get('fileFormat'),
+            OutputLocation=f's3://{environment_bucket}/athenaqueries/{athena_workgroup}/',
+            region=region,
+            AwsAccountId=aws_account_id,
+        )
+        return sql_query_result
+
+    @staticmethod
+    def build_s3_file_path(workgroup: str, query_id: str, athena_queries_dir: str = None) -> str:
+        athena_queries_dir = athena_queries_dir or WorksheetQueryResultService._DEFAULT_ATHENA_QUERIES_PATH
+        return f'{athena_queries_dir}/{workgroup}/{query_id}'
+
+    @staticmethod
+    def convert_csv_to_xlsx(csv_data) -> io.BytesIO:
+        wb = Workbook()
+        ws: 'Worksheet' = wb.active
+        csv_reader = csv.reader(csv_data.splitlines())
+        for row in csv_reader:
+            ws.append(row)
+
+        excel_buffer = io.BytesIO()
+        wb.save(excel_buffer)
+        excel_buffer.seek(0)
+        return excel_buffer
+
+    @staticmethod
+    @TenantPolicyService.has_tenant_permission(RUN_ATHENA_QUERY_TENANT)
```

Comment:
Tenant permissions are generic permissions to restrict a complete module permissions. In this case we are defining permissions for a very particular functionality that affects a single worksheet.


We can use `MANAGE_WORKSHEETS` to restrict/grant access at the tenant level in a very generic way. And then add a resource policy `DOWNLOAD_ATHENA_QUERY_RESULT` that protects the worksheet. We need to modify the signature of the function and define the `uri` input, which should correspond to the worksheet_uri, which no longer will be passed in `data`.

```
@TenantPolicyService.has_tenant_permission(MANAGE_WORKSHEETS)
@ResourcePolicyService.has_resource_permission(DOWNLOAD_ATHENA_QUERY_RESULT)
def download_sql_query_result(session: 'Session', uri: str, env_uri: str, data: dict = None):
```

Reply:
done